// angular-redux-middleware/webpack.config.js
'use strict'

const path = require('path')
const autoprefixer = require('autoprefixer')
const webpack = require('webpack')

const paths = {
  output: __dirname + '/build',
  src: __dirname + '/src',
  nodeModules: 'node_modules',
  entry: './src/index.ts',
  filename: 'index.js',
}

const env = null

module.exports = {
  devtool: 'eval-source-map',
  entry: paths.entry,
  output: {
    path: paths.output,
    filename: paths.filename,
    library: 'angular-redux-middleware',
    libraryTarget: 'umd',
  },
  resolve: {
    extensions: ['.ts', '.json', '.js']
  },
  module: {
    // noParse: [/node_modules\/object-hash\/dist\/object_hash\.js/],
    rules: [{
        exclude: [
          /\.html$/,
          /\.(js)(\?.*)?$/,
          /\.ts/,
          /\.css$/,
          /\.json$/,
          /\.svg$/,
          /\.less$/,
        ],
        use: [{
          loader: 'url-loader',
          options: {
            limit: 10000,
            fallback: 'file-loader',
            name: '/static/media/[name].[hash:8].[ext]',
          },
        }, ],
      },
      {
        test: /\.(js)$/,
        include: paths.src,
        use: [{
          loader: 'babel-loader',
          options: {
            cacheDirectory: true,
            presets: ['angular', 'react-app'],
          },
        }, ],
      },
      {
        test: /\.json$/,
        use: [{
          loader: 'json-loader'
        }],
      },
      {
        test: /\.ts$/,
        include: paths.src,
        loader: 'ts-loader',
      },
    ],
  },
  plugins: [
    new webpack.HotModuleReplacementPlugin(),
  ],
  node: {
    fs: 'empty',
    net: 'empty',
    tls: 'empty',
  },
}
